python/pacemaker/_cts/tests/reattach.py

```diff
@@ -61,23 +62,63 @@ def _set_managed(self, node):
         self.debug("Re-enable resource management")
         self._rsh(node, "crm_attribute -t rsc_defaults -n is-managed -D")
 
+    def _disable_incompatible_rscs(self, node):
+        """ Disable resources that are incompatible with this test
+
+        Starts and stops of stonith-class resources are implemented internally
+        by Pacemaker, which means that they must stop when Pacemaker is
+        stopped, even if unmanaged. Disable them before running the Reattach
+        test so they don't affect resource placement.
+
+        OCFS2 resources must be disabled too for some reason.
+
+        Set target-role to "Stopped" for any of these resources in the CIB.
+        """
+
+        self.debug("Disable incompatible (stonith/OCFS2) resources")
+        xml = """'<meta_attributes id="cts-lab-Reattach-meta">
+                    <nvpair id="cts-lab-Reattach-target-role" name="target-role" value="Stopped"/>
+                    <rule id="cts-lab-Reattach-rule" boolean-op="or" score="INFINITY">
+                      <rsc_expression id="cts-lab-Reattach-stonith" class="stonith"/>
+                      <rsc_expression id="cts-lab-Reattach-o2cb" type="o2cb"/>
+                    </rule>
+                  </meta_attributes>' --scope rsc_defaults"""
+        return self._rsh(node, self._cm.templates['CibAddXml'] % xml)
+
+    def _enable_incompatible_rscs(self, node):
+        """ Re-enable resources that were incompatible with this test """
+
+        self.debug("Re-enable incompatible (stonith/OCFS2) resources")
+        xml = """<meta_attributes id="cts-lab-Reattach-meta">"""
+        return self._rsh(node, """cibadmin --delete --xml-text '%s'""" % xml)
+
+    def _reprobe(self, node):
+        """ Reprobe all resources
+
+        The placement of some resources (such as promotable-1 in the
+        lab-generated CIB) is affected by constraints using node-attribute-based
+        rules. An earlier test may have erased the relevant node attribute, so
+        do a reprobe, which should add the attribute back.
+        """
+
+        return self._rsh(node, """crm_resource --refresh""")
```

Comment:
If you're only ever going to care about the return values of these functions, you could `return self._rsh(...)[0]`.

Reply:
I could see this being more broadly useful in the future. I'm thinking eventually we could have a node class with methods for commands like this, so we're not using `_rsh` directly anywhere but the node methods.